### PR TITLE
Make deriving strategies explicit

### LINKED ISSUE #2479: Compile with -Wmissing-deriving-strategies
The option `-Wmissing-deriving-strategies` (from GHC >= 8.8) enables a warning where a `deriving` clause does not give an explicit strategy. This is considered a best-practice for using deriving strategies, so we should enable it in the project.

### PR DESCRIPTION
**Pull request by ana-pantilie**
Fixes #2479

---

###### Review checklist

The author performs the actions on the checklist. The reviewer evaluates the work and checks the boxes as they are completed.

-   [ ] **Summary.** Write a summary of the changes. Explain what you did to fix the issue, and why you did it. Present the changes in a logical order. Instead of writing a summary in the pull request, you may push a clean Git history.
-   [ ] **Documentation.** Write documentation for new functions. Update documentation for functions that changed, or complete documentation where it is missing.
-   [ ] **Tests.** Write unit tests for every change. Write the unit tests that were missing before the changes. Include any examples from the reported issue as integration tests.
-   [ ] **Clean up.** The changes are already clean. Clean up anything near the changes that you noticed while working. This does not mean only spatially near the changes, but logically near: any code that interacts with the changes!

### CODE REVIEW

File: kore/src/Kore/Attribute/Overload.hs [hunk 1]
```diff
@@ -19,11 +19,13 @@ import Kore.Attribute.Parser as Parser
...
 
+-- TODO: deriving stock (Functor) ???
+
 -- | @Overload@ represents the @overload@ attribute for symbols.
 newtype Overload symbol = Overload {getOverload :: Maybe (symbol, symbol)}
-    deriving (Eq, Ord, Show)
-    deriving (Functor)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Functor)
```
Comment (ana-pantilie) [inline kore/src/Kore/Attribute/Overload.hs:27]: Using the `newtype` strategy here gives:
```
    • Can't make a derived instance of
        ‘Functor Overload’ with the newtype strategy:
        cannot eta-reduce the representation type enough
```

File: kore/src/Kore/Internal/InternalMap.hs [hunk 2]
```diff
@@ -45,26 +45,26 @@ type MapElement = Element NormalizedMap
 
 -- | Wrapper for normalized maps, to be used in the `builtinAcChild` field.
 newtype NormalizedMap key child = NormalizedMap {getNormalizedMap :: NormalizedAc NormalizedMap key child}
-    deriving (Eq, Ord, Show)
-    deriving (Foldable, Functor, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Foldable, Functor, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Internal/InternalMap.hs:49]: When `deriving newtype (Traversable)`:
```
    • Couldn't match representation of type ‘m (NormalizedAc
                                                  NormalizedMap key a)’
                               with that of ‘m (NormalizedMap key a)’
        arising from the coercion of the method ‘sequence’
          from type ‘forall (m :: * -> *) a.
                     Monad m =>
                     NormalizedAc NormalizedMap key (m a)
                     -> m (NormalizedAc NormalizedMap key a)’
            to type ‘forall (m :: * -> *) a.
                     Monad m =>
                     NormalizedMap key (m a) -> m (NormalizedMap key a)’
      NB: We cannot know what roles the parameters to ‘m’ have;
        we must assume that the role is nominal
    • When deriving the instance for (Traversable (NormalizedMap key))
```

File: kore/src/Kore/Internal/InternalMap.hs [hunk 3]
```diff
@@ -45,26 +45,26 @@ type MapElement = Element NormalizedMap
...
     deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
     deriving anyclass (Debug, Diff)
 
 instance AcWrapper NormalizedMap where
     newtype Value NormalizedMap child = MapValue {getMapValue :: child}
-        deriving (Eq, Ord, Show)
-        deriving (Foldable, Functor, Traversable)
-        deriving (GHC.Generic)
+        deriving stock (Eq, Ord, Show)
+        deriving stock (Foldable, Functor, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Internal/InternalMap.hs:58]: Switching to `newtype` deriving gives:
```
    • No instance for (Traversable (Value NormalizedMap))
        arising from the first field of ‘NormalizedMap’
          (type ‘NormalizedAc NormalizedMap key child’)
      Possible fix:
        use a standalone 'deriving instance' declaration,
          so you can specify the instance context yourself
    • When deriving the instance for (Traversable (NormalizedMap key))
```
The error is the same for `Foldable` and `Functor`.

Comment (ana-pantilie) [inline kore/src/Kore/Internal/InternalMap.hs:58]: Switching to newtype deriving: the actual error seems to be the `cannot eta-reduce` one.

File: kore/src/Kore/Internal/InternalMap.hs [hunk 4]
```diff
@@ -45,26 +45,26 @@ type MapElement = Element NormalizedMap
...
         deriving anyclass (Hashable, NFData)
         deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
         deriving anyclass (Debug, Diff)
 
     newtype Element NormalizedMap child = MapElement {getMapElement :: (child, child)}
-        deriving (Eq, Ord, Show)
-        deriving (Foldable, Functor, Traversable)
-        deriving (GHC.Generic)
+        deriving stock (Eq, Ord, Show)
+        deriving stock (Foldable, Functor, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Internal/InternalMap.hs:66]: Switching to newtype deriving: the actual error seems to be the `cannot eta-reduce` one.

File: kore/src/Kore/Internal/InternalSet.hs [hunk 5]
```diff
@@ -48,26 +48,26 @@ type SetElement = Element NormalizedSet
 
 -- | Wrapper for normalized sets, to be used in the `builtinAcChild` field.
 newtype NormalizedSet key child = NormalizedSet {getNormalizedSet :: NormalizedAc NormalizedSet key child}
-    deriving (Eq, Ord, Show)
-    deriving (Foldable, Functor, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Foldable, Functor, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Internal/InternalSet.hs:52]: See `NormalizedMap`.

File: kore/src/Kore/Internal/InternalSet.hs [hunk 6]
```diff
@@ -48,26 +48,26 @@ type SetElement = Element NormalizedSet
...
         deriving anyclass (Hashable, NFData)
         deriving anyclass (SOP.Generic, SOP.HasDatatypeInfo)
         deriving anyclass (Debug, Diff)
 
     newtype Element NormalizedSet child = SetElement {getSetElement :: child}
-        deriving (Eq, Ord, Show)
-        deriving (Foldable, Functor, Traversable)
-        deriving (GHC.Generic)
+        deriving stock (Eq, Ord, Show)
+        deriving stock (Foldable, Functor, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Internal/InternalSet.hs:69]: See `MapElement`.

File: kore/src/Kore/Syntax/Bottom.hs [hunk 7]
```diff
@@ -23,9 +23,9 @@ category from the Semantics of K, Section 9.1.4 (Patterns).
 'bottomSort' is the sort of the result.
 -}
 newtype Bottom sort child = Bottom {bottomSort :: sort}
-    deriving (Eq, Ord, Show)
-    deriving (Functor, Foldable, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Functor, Foldable, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Syntax/Bottom.hs:27]: Same `eta-reduce` error.

File: kore/src/Kore/Syntax/Inhabitant.hs [hunk 8]
```diff
@@ -22,9 +22,9 @@ import Prelude.Kore
 
 -- | 'Inhabitant' symbolizes the inhabitants of a sort.
 newtype Inhabitant child = Inhabitant {inhSort :: Sort}
-    deriving (Eq, Ord, Show)
-    deriving (Functor, Foldable, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Functor, Foldable, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Syntax/Inhabitant.hs:26]: `eta-reduce` error.

File: kore/src/Kore/Syntax/Pattern.hs [hunk 9]
```diff
@@ -98,9 +98,9 @@ newtype
         (variable :: Type)
         (annotation :: Type) = Pattern
     {getPattern :: Cofree (PatternF variable) annotation}
-    deriving (Show)
-    deriving (Functor, Foldable, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Show)
+    deriving stock (Functor, Foldable, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Syntax/Pattern.hs:102]: Issue with `Traversable` instance again.

File: kore/src/Kore/Syntax/Sentence.hs [hunk 10]
```diff
@@ -355,9 +355,9 @@ data SentenceAxiom (patternType :: Type) = SentenceAxiom
     , sentenceAxiomPattern :: !patternType
     , sentenceAxiomAttributes :: !Attributes
     }
-    deriving (Eq, Ord, Show)
-    deriving (Functor, Foldable, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Functor, Foldable, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Syntax/Sentence.hs:359]: Issue with `Traversable`.

File: kore/src/Kore/Syntax/Top.hs [hunk 11]
```diff
@@ -22,9 +22,9 @@ import Prelude.Kore
 'topSort' is the sort of the result.
 -}
 newtype Top sort child = Top {topSort :: sort}
-    deriving (Eq, Ord, Show)
-    deriving (Functor, Foldable, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Functor, Foldable, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Syntax/Top.hs:26]: Eta-reduce.

File: kore/src/Kore/Syntax/Variable.hs [hunk 12]
```diff
@@ -170,9 +170,9 @@ instance From VariableName VariableName where
 -- * Element variables
 
 newtype ElementVariableName variable = ElementVariableName {unElementVariableName :: variable}
-    deriving (Eq, Ord, Show)
-    deriving (Functor, Foldable, Traversable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Functor, Foldable, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Syntax/Variable.hs:174]: `Eta-reduce`.

File: kore/src/Kore/Syntax/Variable.hs [hunk 13]
```diff
@@ -213,10 +213,10 @@ instance
     from = from . unElementVariableName
 
 newtype SetVariableName variable = SetVariableName {unSetVariableName :: variable}
-    deriving (Eq, Ord, Show)
-    deriving (Functor)
-    deriving (Foldable, Traversable)
-    deriving (GHC.Generic, GHC.Generic1)
+    deriving stock (Eq, Ord, Show)
+    deriving stock (Functor)
+    deriving stock (Foldable, Traversable)
```
Comment (ana-pantilie) [inline kore/src/Kore/Syntax/Variable.hs:218]: `Eta-reduce`.

File: kore/src/SQL/Key.hs [hunk 14]
```diff
@@ -18,8 +18,8 @@ import Prelude.Kore
 
 -- | A foreign key into the table for type @a@.
 newtype Key a = Key {getKey :: Int64}
-    deriving (Eq, Ord, Read, Show)
-    deriving (Functor, Foldable)
-    deriving (GHC.Generic)
+    deriving stock (Eq, Ord, Read, Show)
+    deriving stock (Functor, Foldable)
```
Comment (ana-pantilie) [inline kore/src/SQL/Key.hs:22]: `Eta-reduce`.

File: kore/src/Kore/Log/InfoExecDepth.hs [hunk 15]
```diff
@@ -20,16 +20,16 @@ import Pretty (
 import qualified Pretty
 
 newtype ExecDepth = ExecDepth {getExecDepth :: Natural}
-    deriving (Eq, Ord, Show)
-    deriving (Enum)
+    deriving stock (Eq, Ord, Show)
+    deriving newtype (Enum)
```
Comment (ana-pantilie) [inline kore/src/Kore/Log/InfoExecDepth.hs:24]: Actually, I'm not sure if this should be `stock` or `newtype`.

File: kore/src/Kore/Log/InfoProofDepth.hs [hunk 16]
```diff
@@ -21,8 +21,8 @@ import Pretty (
 import qualified Pretty
 
 newtype ProofDepth = ProofDepth {getProofDepth :: Natural}
-    deriving (Eq, Ord, Show)
-    deriving (Enum)
+    deriving stock (Eq, Ord, Show)
+    deriving newtype (Enum)
```
Comment (ana-pantilie) [inline kore/src/Kore/Log/InfoProofDepth.hs:25]: Same question as above.

**Review by ttuegel — APPROVED** [review]

**Pull request merged**